opsimulator/opsimulator.go

```diff
@@ -216,9 +257,120 @@ func (opSim *OpSimulator) AddDependency(chainID uint64, depSet []uint64) error {
 	return nil
 }
 
-// TODO(https://github.com/ethereum-optimism/supersim/issues/19): add logic for checking that an interop transaction is valid.
-func checkInteropInvariants() bool {
-	return true
+func (opSim *OpSimulator) checkInteropInvariants(ctx context.Context, tx *types.Transaction) error {
+	from, err := getFromAddress(tx)
+
+	if err != nil {
+		return fmt.Errorf("failed to find sender of transaction: %w", err)
+	}
+
+	result, err := opSim.l2Chain.DebugTraceCall(ctx, config.TransactionArgs{From: from, To: tx.To(), Gas: hexutil.Uint64(tx.Gas()), GasPrice: (*hexutil.Big)(tx.GasPrice()), Data: tx.Data(), Value: (*hexutil.Big)(tx.Value())})
+	if err != nil {
+		return fmt.Errorf("failed to simulate transaction: %w", err)
+	}
+	if result.Error != nil {
+		return fmt.Errorf("tx trace error: %v", result.Error)
+	}
+	simulatedLogs := toSimulatedLogs(result)
+
+	crossL2Inbox := crossl2inbox.NewCrossL2Inbox()
+	var executingMessages []crossl2inbox.ExecutingMessage
+	for _, log := range simulatedLogs {
+		executingMessage, err := crossL2Inbox.DecodeExecutingMessageLog(&log)
+		if err != nil {
+			return fmt.Errorf("failed to decode executing messages from transaction logs: %w", err)
+		}
+
+		if executingMessage != nil {
+			executingMessages = append(executingMessages, *executingMessage)
+		}
+	}
+
+	if len(executingMessages) >= 1 {
+		for _, executingMessage := range executingMessages {
+			client := opSim.clients[executingMessage.Identifier.ChainId.Uint64()]
+			if client == nil {
+				return fmt.Errorf("no client found for chainid: %v on executing message", executingMessage.Identifier.ChainId)
+			}
+
+			block, err := client.EthBlockByNumber(ctx, executingMessage.Identifier.BlockNumber)
+			if err != nil {
+				return fmt.Errorf("failed to fetch executing message block: %w", err)
+			}
+
+			if executingMessage.Identifier.Timestamp.Cmp(new(big.Int).SetUint64(block.Time())) != 0 {
+				return fmt.Errorf("executing message identifier does not match block timestamp: %w", err)
+			}
+
+			logs, err := client.EthGetLogs(
+				ctx,
+				ethereum.FilterQuery{
+					Addresses: []common.Address{executingMessage.Identifier.Origin},
+					FromBlock: executingMessage.Identifier.BlockNumber,
+					ToBlock:   executingMessage.Identifier.BlockNumber,
+				},
+			)
+			if err != nil {
+				return fmt.Errorf("failed to fetch initiating message logs: %w", err)
+			}
+			var initiatingMessageLogs []ethTypes.Log
```

Comment:
since we retrieved the logs using eth_getLogs is it guaranteed that the array of logs returned from that rpc request match with logIndex? According to the spec logIndex is the hexadecimal of the log index position in the block.

Reply:
ah good point, I think in my backend impl, i'd fetch all the logs in the block. Not filtered by contract. Makes sense!